frontend/catalyst/qjit_device.py

```diff
@@ -218,3 +220,112 @@ def _decomp_to_unitary(self, *_args, **_kwargs):
+        return config["compilation"]["mid_circuit_measurement"]
+
+    @staticmethod
+    def _check_adjoint(config):
+        return config["compilation"]["quantum_adjoint"]
+
+    @staticmethod
+    def _check_quantum_control(config):
+        return config["compilation"]["quantum_control"]
+
+    @staticmethod
+    def _set_supported_operations(config):
+        """Override the set of supported operations."""
+        native_gates = set(config["operators"]["gates"][0]["native"])
+        qir_gates = QJITDeviceNewAPI.operations_supported_by_QIR_runtime
+        QJITDeviceNewAPI.operations = list(native_gates.intersection(qir_gates))
+
+        # These are added unconditionally.
+        QJITDeviceNewAPI.operations += ["Cond", "WhileLoop", "ForLoop"]
+
+        if QJITDeviceNewAPI._check_mid_circuit_measurement(config):  # pragma: no branch
+            QJITDeviceNewAPI.operations += ["MidCircuitMeasure"]
+
+        if QJITDeviceNewAPI._check_adjoint(config):
+            QJITDeviceNewAPI.operations += ["Adjoint"]
+
+        if QJITDeviceNewAPI._check_quantum_control(config):  # pragma: nocover
+            # TODO: Once control is added on the frontend.
+            gates_to_be_decomposed_if_controlled = [
+                "Identity",
+                "CNOT",
+                "CY",
+                "CZ",
+                "CSWAP",
+                "CRX",
+                "CRY",
+                "CRZ",
+                "CRot",
+            ]
+            native_controlled_gates = ["ControlledQubitUnitary"] + [
+                f"C({gate})"
+                for gate in native_gates
+                if gate not in gates_to_be_decomposed_if_controlled
+            ]
+            QJITDevice.operations += native_controlled_gates
+
+    @staticmethod
+    def _set_supported_observables(config):
+        """Override the set of supported observables."""
+        QJITDeviceNewAPI.observables = config["operators"]["observables"]
+
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self,
+        original_device,
+        config,
+        backend_name=None,
+        backend_lib=None,
+        backend_kwargs=None,
+    ):
+        self.original_device = original_device
+
+        for key, value in original_device.__dict__.items():
+            self.__setattr__(key, value)
+        self.config = config
+        QJITDeviceNewAPI._set_supported_operations(self.config)
+        QJITDeviceNewAPI._set_supported_observables(self.config)
+
+        self.backend_name = backend_name if backend_name else "default"
+        self.backend_lib = backend_lib if backend_lib else ""
+        self.backend_kwargs = backend_kwargs if backend_kwargs else {}
+
+    def preprocess(
+        self,
+        execution_config: qml.devices.ExecutionConfig = qml.devices.DefaultExecutionConfig,
+    ):
+        """Device preprocessing function."""
+        program, config = self.original_device.preprocess(execution_config)
+        return program, config
```

Comment:
I don't see how the TOML properties are taken into account by this new device 🤔

Is the strategy to just trust the device pre-processes according to it's TOML (reasonable), but then it doesn't take into account the Gate set limitations in the runtime, or Catalyst specific operations like control flow etc?

Reply:
I don't think it was taken into account in the previous device as well and that's why we have a goal to validate programs.

For this device we can add couple of TOML based transforms here.

Reply:
Can you add a TODO here that we need to add decompositions for the parsed operations? :)

Reply:
Done, thanks!